File: first-bad-version/solution278.go
// https://leetcode.com/problems/first-bad-version/

package main

import "fmt"

func isBadVersion(n int) bool {
	return n >= 4
}

// T: O(log n), S: O(1)
func firstBadVersion(n int) int {
	left, right := 1, n
	for left < right {
		mid := left + (right-left)/2
		if isBadVersion(mid) {
			right = mid
		} else {
			left = mid + 1
		}
	}
	return left
}

func main() {
	fmt.Println(firstBadVersion(5)) // 4
}

File: first-bad-version/Solution278.java
// https://leetcode.com/problems/first-bad-version/

class VersionControl {
    boolean isBadVersion(int n) {
        return n >= 4;
    }
}

class Solution278 extends VersionControl {

    // T: O(log n), S: O(1)
    public int firstBadVersion(int n) {
        int left = 1, right = n;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (isBadVersion(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    public static void main(String[] args) {
        var sol = new Solution278();
        System.out.println(sol.firstBadVersion(5)); // 4
    }

}
